README.md

```diff
@@ -278,3 +278,40 @@ jsonPayload.verb="stacktraceanalyzer"
+
+## Cross Ahead of Time Compilation (XAOT)
+To compile against target hardware ahead of time, we provide a tool `train_compile.py`. This tool allows you to compile the main `train_step` in `train.py` for target hardware (e.g. a large number of v5e devices) without using the target hardware, and instead you may use only a single VM from a different family, e.g. a v4-8. This compilation helps with two main goals:
+
+* It will flag any out of memory (OOM) information, such as when the `per_device_batch_size` is set too high, with an identical OOM stack trace as if it was compiled on the target hardware.
+
+* The ahead of time compilation can be saved and then loaded for fast startup and restart times on the target hardware.
+
+The tool `train_compile.py` is tightly linked to `train.py` and uses the same configuration file `configs/base.yml`. Here is an example run:
+```
+# Run the below on a single machine, e.g. a v4-8
+python3 MaxText/train_compile.py MaxText/configs/base.yml compile_topology=v5e-256 compile_topology_num_slices=2 \ 
+global_parameter_scale=16 per_device_batch_size=4
+```
+
+This will compile a 16B parameter MaxText model on 2 v5e pods. 
+
+Here is an example that saves then loads the compiled `train_step`, starting with the save:
+```
+# Run the below on a single machine, e.g. a v4-8
+export LIBTPU_INIT_ARGS="--xla_enable_async_all_gather=true" 
+python3 MaxText/train_compile.py MaxText/configs/base.yml compile_topology=v5e-256 \
+compile_topology_num_slices=2 \
+compile_save_file=my_compiled_train.pickle global_parameter_scale=16 \
+per_device_batch_size=4 steps=10000 learning_rate=1e-3
+```
+
+To load the compiled train_step, you just need to pass `compile_save_file=my_compiled_train.pickle` into `train.py`:
+```
+# Run the below on each host of the target hardware, e.g. each host on 2 slices of v5e-256
+python3 MaxText/train.py MaxText/configs/base.yml run_name=example_load_compile \
+compile_save_file=my_compiled_train.pickle \
+global_parameter_scale=16  per_device_batch_size=4 steps=10000 \
+base_output_directory=gs://my-output-bucket dataset_path=gs://my-dataset-bucket
+```
+
+Above we included exporting the compiler flag `LIBTPU_INIT_ARGS` and `learning_rate` because those affect the compiled object `my_compiled_train.pickle.` The sizes of the model (e.g. `global_parameter_scale`, `max_sequence_length` and `per_device_batch`) are fixed when you initally compile via `compile_train.py`, you will see a size error if you try to run the saved compiled object with different sizes than you compiled with. However a subtle note is that the **learning rate schedule** is also fixed when you run `compile_train` - which is determined by both `steps` and `learning_rate`. The optimizier parameters such as  `adam_b1` are passed only as shaped objects to the compiler - thus their real values are determined when you run `train.py`, not during the compilation.
```

Comment:
Gotta have some sort of communication about what the failure mode is here!

Reply:
Added notes on expected failure modes, lmk what you think